ragna/ui/components/file_uploader.py

```diff
@@ -0,0 +1,249 @@
+import json
+
+import param
+from panel.reactive import ReactiveHTML
+from panel.widgets import Widget
+
+
+class FileUploader(ReactiveHTML, Widget):  # type: ignore[misc]
+    file_list = param.List(default=[])
+
+    custom_js = param.String(default="")
+    uploaded_documents_json = param.String(default="")
+
+    title = param.String(default="")
+
+    def __init__(self, token, informations_endpoint, **params):
+        super().__init__(**params)
+
+        self.token = token
+        self.informations_endpoint = informations_endpoint
+
+        self.after_upload_callback = None
+
+    @param.depends("uploaded_documents_json", watch=True)
+    def did_finish_upload(self):
+        if self.after_upload_callback is not None:
+            self.after_upload_callback(json.loads(self.uploaded_documents_json))
+
+    def perform_upload(self, event=None, after_upload_callback=None):
+        self.after_upload_callback = after_upload_callback
+
+        self.loading = True
+
+        final_callback_js = """
+            var final_callback = function(uploaded_documents) {
+                self.get_uploaded_documents_json().innerText = JSON.stringify(uploaded_documents);
+            };
+        """
+
+        self.custom_js = (
+            final_callback_js
+            + f""" upload( self.get_upload_files(),  '{self.token}', '{self.informations_endpoint}', final_callback) """
+        )
+
+    _child_config = {
+        "custom_js": "template",
+        "uploaded_documents_json": "template",
+    }
+
+    _template = """
```

Comment:
Can we please move all JS / CSS that is longer than say 3 lines to a `.js` / `.css` file? This module is really hard to read.

Reply:
Yes that's possible, that's why we have `upload.js` for instance. But I'll do that later, the things are still moving a lot so for the moment it's more convenient for me to have everything at one place than scattered.

It's hard to read because that's a custom Panel component, so it holds 3 parts **in multiple languages** : 
- its HTML template, which includes its CSS styling
- all the JS machinery to make it work in the frontend, and junction with Python
- the python part, which is the backend

The CSS will be moved somewhere else and simplified anyway, the JS part can move in a dedicated file.